### DISCUSSION TITLE
Any way to use `id` field instead of `_id` with mongo driver?

### DISCUSSION THREAD

**Discussion by paweltatarczuk** [mikro-orm/mikro-orm#3171] [Q&A] [answered]
It seems that mongo driver is rewriting `id` field to `_id` automatically:
https://github.com/mikro-orm/mikro-orm/blob/master/packages/mongodb/src/MongoDriver.ts#L108

Is there any way to query entities by the `id` field?

Sample entity:

```
import { EntitySchema } from '@mikro-orm/core';

export interface Foo = {
  _id: string,
  id: string,
  name: string,
};

export const schema = new EntitySchema<Foo>({
  name: 'Foo',
  properties: {
    _id: { type: 'objectid', primary: true, hidden: true },
    id: { type: 'string' },
    name: { type: 'string' },
  },
});
```

**Comment by B4nan** [comment, marked as answer]
Your schema is wrong, you define the `id` as regular property, which has its own value. Mark is as `serializedPrimaryKey: true`, this way it will mirror the `_id` value:

```ts
export const schema = new EntitySchema<Foo>({
  name: 'Foo',
  properties: {
    _id: { type: 'objectid', primary: true }, // no need for explicit hidden: true, the serialized PK is always preferred
    id: { type: 'string', serializedPrimaryKey: true },
    name: { type: 'string' },
  },
});
```

**Reply by paweltatarczuk** [reply]
The thing is I don't want the `id` field to mirror the `_id` value. For example:

```
{
	"_id" : ObjectId("625e4a101c07bfc2e866c794"),
	"id" : "MvMsPbFLBZ",
	"name" : "Bar",
}
```

When I query for the ID field like so I get an empty result:

```
const bar = repo.find({ id: 'MvMsPbFLBZ' }); // undefined
```

**Reply by B4nan** [reply]
Oh I see. Is that particular line the only issue, have you tried commenting it out?

**Reply by B4nan** [reply]
As a workaround, you could query via the mongo collection directly and map to entities afterwards:

```ts
const res = await em.getCollection(Ent).find(...);
const ents = res.map(row => em.map(Ent, row));
```

**Reply by paweltatarczuk** [reply]
After commenting this line out I'm able to use `id` field as any other field. I'm not sure if everything works properly now but for now all seems fine (I'm using only some basic stuff now, no relations etc.).

For now I can patch the driver and try moving forward. Maybe this behavior could be configurable? I could open up a PR if this change is not breaking anything. If not I would probably move away from the `id` field in future (which wasn't a good choice back then anyway).

**Reply by B4nan** [reply]
It should work based on the metadata (based on the `serializedPrimaryKey` option), that line is a relict of v1 basically, where things were not configurable at all and your use case was not supported. The problem is that fixing this might actually be breaking, someone might be depending on this behaviour, in such case it would have to be under a feature toggle before v6 is out.

**Reply by paweltatarczuk** [reply]
Thank you! I'll post a PR if I can come up with anything helpful!

**Reply by paweltatarczuk** [reply]
I'm marking this as answered, I hope it's okay that the answer is in the thread comments

**Reply by paweltatarczuk** [reply]
I'm attaching a patch that is working so far for my case, there were more changes required to be able to find entities by `id` than just the one discussed.
[@mikro-orm+mongodb+^5.1.4.patch.txt](https://github.com/mikro-orm/mikro-orm/files/8840726/%40mikro-orm%2Bmongodb%2B.5.1.4.patch.txt)

Looks like this could be configurable by the entity's metadata (the current behavior is hard-coded, `_id` has to be the primary key and `id` has to be serialized primary key no matter the entity's metadata),  but as it was said before , it would be a breaking change.

**Reply by B4nan** [reply]
For some reason I can't open that file, can you past the contents here directly?

**Reply by paweltatarczuk** [reply]
Sure:

```
# generated by patch-package 6.4.14
#
# declared package:
#   @mikro-orm/mongodb: ^5.1.4
#
diff --git a/node_modules/@mikro-orm/mongodb/MongoDriver.js b/node_modules/@mikro-orm/mongodb/MongoDriver.js
index 13f2f0b..5f4d89e 100644
--- a/node_modules/@mikro-orm/mongodb/MongoDriver.js
+++ b/node_modules/@mikro-orm/mongodb/MongoDriver.js
@@ -42,7 +42,7 @@ class MongoDriver extends core_1.DatabaseDriver {
         data = data.map(d => this.renameFields(entityName, d));
         const meta = this.metadata.find(entityName);
         /* istanbul ignore next */
-        const pk = meta?.getPrimaryProps()[0].fieldNames[0] ?? '_id';
+        const pk = meta?.getPrimaryProps()[0].fieldNames[0] ?? 'id';
         const res = await this.rethrow(this.getConnection('write').insertMany(entityName, data, options.ctx));
         res.rows = res.insertedIds.map(id => ({ [pk]: id }));
         return res;
@@ -74,7 +74,6 @@ class MongoDriver extends core_1.DatabaseDriver {
     }
     renameFields(entityName, data, where = false) {
         data = Object.assign({}, data); // copy first
-        core_1.Utils.renameKey(data, 'id', '_id');
         const meta = this.metadata.find(entityName);
         if (meta) {
             this.inlineEmbeddables(meta, data, where);
@@ -136,7 +135,7 @@ class MongoDriver extends core_1.DatabaseDriver {
         if (meta.properties[meta.primaryKeys[0]].type.toLowerCase() === 'objectid') {
             return { _id: new bson_1.ObjectId(id) };
         }
-        return { _id: id };
+        return { id };
     }
     buildFields(entityName, populate, fields) {
         const meta = this.metadata.find(entityName);
diff --git a/node_modules/@mikro-orm/mongodb/MongoPlatform.js b/node_modules/@mikro-orm/mongodb/MongoPlatform.js
index 13345f4..7ee1cc5 100644
--- a/node_modules/@mikro-orm/mongodb/MongoPlatform.js
+++ b/node_modules/@mikro-orm/mongodb/MongoPlatform.js
@@ -30,10 +30,10 @@ class MongoPlatform extends core_1.Platform {
         return new bson_1.ObjectId(data);
     }
     getSerializedPrimaryKeyField(field) {
-        return 'id';
+        return field;
     }
     usesDifferentSerializedPrimaryKey() {
-        return true;
+        return false;
     }
     usesImplicitTransactions() {
         return false;
@@ -56,10 +56,7 @@ class MongoPlatform extends core_1.Platform {
         return prop.reference === core_1.ReferenceType.MANY_TO_MANY && prop.owner;
     }
     validateMetadata(meta) {
-        const pk = meta.getPrimaryProps()[0];
-        if (pk && pk.fieldNames?.[0] !== '_id') {
-            throw core_1.MetadataError.invalidPrimaryKey(meta, pk, '_id');
-        }
+      return;
     }
 }
 exports.MongoPlatform = MongoPlatform;
```
